<div style="border:solid Chocolate 2px; padding: 40px">

Привет, меня зовут Александр Куимов. Я буду ревьюером твоего проекта. Ты можешь обращаться ко мне на "ты"😏 Надеюсь, тебя также не смутит, если я буду обращаться к тебе на "ты", но если это неудобно, обязательно скажи об этом!

Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 

Ты также можешь реагировать на мои комментарии своими по шаблону, показанному чуть ниже. Это нужно, чтобы не создавалась путаница😉

Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
    
<br/>    
<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>👋:</b> В такой цветовой ячейке я прошу тебя оставлять свои комментарии. Если исправляешь проект на второй итерации и выше, не забывай пожалуйста указывать номер итерации, например, "Комментарий студента v.2".
</div> 

<br/>    
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также, и я пытаюсь подготовить тебя именно к работе датасаентистом. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Вижу твое добавленное описание проекта. Молодец! Это поможет тебе расставлять акценты в выводах

In [1]:
import pandas as pd
import numpy as np

import re 
import nltk
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('averaged_perceptron_tagger')
from nltk.stem import WordNetLemmatizer
from nltk import pos_tag, word_tokenize
from nltk.corpus import wordnet
from nltk.corpus import stopwords as nltk_stopwords

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score
RANDOM_STATE = 42
TEST_SIZE = 0.25

from tqdm import tqdm # для progress_apply
tqdm.pandas()         # для progress_apply

[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/jovyan/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


## Подготовка

In [2]:
data = pd.read_csv('/datasets/toxic_comments.csv')

In [3]:
data.info()
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

Пропущенных значений нет, типы данных указаны верно.

Займемся лемматизацией текста:

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>  
    
Напоминаю о важности исследования баланса классов. Заранее предупрежу, что по условию задания балансировать нам не обязательно, а если применять кросс-валидацию к сбалансированной обучающей выборке, то это будет ошибкой, поэтому понадобится балансирование с помощью готовых решений (class_weight, SMOTE, Tomek Links…) 
        

In [4]:
#def lemmatize(text):
#    lem = WordNetLemmatizer()
#    return ' '.join([lem.lemmatize(w) for w in text.split()])
#def clear_text(text):
#    text = re.sub(r'[^a-zA-Z ]', ' ', text)
#    text = text.split()
#    return ' '.join(text)

wnl = WordNetLemmatizer()

def penn2morphy(penntag):
    morphy_tag = {'NN':'n', 'JJ':'a',
                  'VB':'v', 'RB':'r'}
    try:
        return morphy_tag[penntag[:2]]
    except:
        return 'n' 

def lemmatize_sent(text):
    text = re.sub(r'[^a-zA-Z ]', ' ', text)
    return ' '.join([wnl.lemmatize(word.lower(), pos=penn2morphy(tag)) 
            for word, tag in pos_tag(word_tokenize(text))])

In [5]:
data['lemm_text'] = data['text'].progress_apply(lemmatize_sent)

100%|██████████| 159292/159292 [09:59<00:00, 265.88it/s]


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера   <a class="tocSkip"></h2>
    
<b>На доработку🤔:</b>
   
    
Нужно применить WordNetLemmatizer() **(+POS тег)** или SpaCy. 
    
Пример кода можно взять из тренажера. **Обрати внимание, что лемматизацию мы должны применять к каждому токену (слову в тексте). made в нулевой строке должно стать make, а are в четвертой строке стать be**
    
    
Делюсь с тобой полезными ссылками:
        
  
 https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/
        
Хочу обратить твое внимание, что код из вышеуказанной статьи выполняется несколько дольше, чем оптимизированный, вот из этого топика
    
https://stackoverflow.com/questions/50992974/nltk-wordnetlemmatizer-not-lemmatizing-as-expected    
        
       
    
    
<div class="alert alert-warning">    
<b>Некоторые замечания и рекомендации💡:</b> 
   


- лемматизацию можно было сделать с помощью SpaCy лемматизатором и прямо скажем как инструмент он более удобен и универсален, не нужно заморачиваться с токенизацией и учётом пос тегов




- Чтобы сэкономить время, и убедиться что всё отработало нормально, берёшь парочку предложений, создаёшь dataframe
    
    
    sentence1 = "The striped bats are hanging on their feet for best"
    sentence2 = "you should be ashamed of yourself went worked"
    df_my = pd.DataFrame([sentence1, sentence2], columns = ['text'])
    print(df_my)


    print(df_my['text'].apply(func))
    
    
    
И тестируешь не нем, должно получиться 
    
    
    
    striped  ------> stripe, went -------> go  



Если всё получилось, то можно использовать на всём датасете

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

Остановился на ускоренном коде WordNetLemmatizer из статьи. Он чуть ли не в 3раза быстрее обычного WordNetLemmatizer.
    
Изначально пытался использовать SpaCy, прочитал о нем, что он популярнее чем WordNetLemmatizer. Но он очень долго выполняется, я оставлял его сегодня на ночь и с утра получил ошибку: maximum recursion depth exceeded in comparison. Код брал из твоей ссылки:
   
    nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])
    doc = nlp(sentence)
    " ".join([token.lemma_ for token in doc])
    
Так же заметил что в ускоренном варианте WordNetLemmatizer больше ошибок, что важнее в реальной работе идеально лемматизированный текст или быстрота кода?
</div> 

 <div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

Важно найти баланс между быстротой кода и качеством лемматизации. Если нужна высокая точность, то лучше использовать решение, которое обеспечивает более качественную лемматизацию, даже если оно выполняется медленнее.
Что касается ошибки “maximum recursion depth exceeded”, то она обычно возникает, когда программа пытается вызвать саму себя рекурсивно слишком много раз. Это может произойти, если есть бесконечный цикл или очень глубокая рекурсия в коде. Чтобы исправить эту проблему, можно увеличить глубину рекурсии, добавив следующий код:
    import sys 
    import os 
    sys.setrecursionlimit(1500)

Это увеличит глубину рекурсии до 1500 уровней

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>
    

    
Можно использовать progress_apply. Для контроля времени лемматизации, поскольку процесс длительный, можно применять индикатор прогресса. Различные варианты можно посмотреть здесь: 
    
https://habr.com/ru/post/483400/ 
    
Пример:
    
    from tqdm import tqdm
    tqdm.pandas()
    data['lemm_text'] = data['text'].progress_apply(lemmatize)

</div>

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>
    
progress_apply очень выручил, есть ли что-то похожее для работы с GridSearchCV при обучении на тренировочный данных? 
   
В этом проекте search.fit(X_train,y_train) такой же долгий процесс как и выделение лемм. Очень сложно заниматься подбором параметров из-за этого.
</div>

 <div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

Для GridSearchCV можно использовать параметр verbose=True, который выводит информацию о ходе выполнения поиска. Однако, это не так удобно, как progress_apply, так как не показывает прогресс в процентах


Разобъем признаки на входные и целевой

In [6]:
X = data['lemm_text']
y = data['toxic']

Разобъем выборку на train и test

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = TEST_SIZE, random_state = RANDOM_STATE)

Избавимся от слов не имеющих смысловой нагрузки и вычислим TF-IDF

In [8]:
stopwords = set(nltk_stopwords.words('english'))
count_tf_idf = TfidfVectorizer(stop_words=stopwords)

In [9]:
X_train_tfidf = count_tf_idf.fit_transform(X_train)
X_test_tfidf = count_tf_idf.transform(X_test)

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>
    
Разбиение сделано верно. Отлично, что векторизатор обучен только на тренировочной части данных
</div>

## Обучение

Будем обучать две модели LogisticRegression и DecisionTreeClassifier, метрикой для проверки будем использовать F1_score

In [10]:
pipe_final = Pipeline([('models', LogisticRegression(random_state = RANDOM_STATE))])

In [11]:
param_distributions = [
    {
        'models': [LogisticRegression(random_state = RANDOM_STATE, penalty='elasticnet', solver='saga')],
        'models__l1_ratio': [0, 0.5, 1],
    },
    {
        'models': [DecisionTreeClassifier(random_state = RANDOM_STATE)],
        'models__max_depth': [2, 5],
        'models__min_samples_leaf': [2, 5]
    }
] 

randomized_search = GridSearchCV(
    pipe_final, 
    param_distributions, 
    scoring='f1', 
    n_jobs=-1
)

In [12]:
randomized_search.fit(X_train_tfidf, y_train);

In [13]:
print('Лучшая модель и её параметры:\n', randomized_search.best_estimator_) 

y_test_pred = randomized_search.predict(X_test_tfidf)
print(f'Метрика RMSE для лучшей модели: {randomized_search.best_score_}')

Лучшая модель и её параметры:
 Pipeline(steps=[('models',
                 LogisticRegression(l1_ratio=1, penalty='elasticnet',
                                    random_state=42, solver='saga'))])
Метрика RMSE для лучшей модели: 0.7586272969149723


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Линейная модель + подбор гиперпараметра С - это секрет успеха в этом проекте. Поэтому есть смысл  поработать с регуляризацией (пеналти l1/l2 + подбор гиперпараметра С в диапазоне 5-15). Это поможет логистической регрессии (или любой другой линейной модели классификации) раскрыться в полную силу
        


Лучшей стала LogisticRegression с метрикой на тренировочных данных равной 0.758.

In [14]:
pd.DataFrame(randomized_search.cv_results_)[['mean_test_score', 'mean_fit_time',
                                                          'mean_score_time', 'param_models', 'params']]

mean_test_score  mean_fit_time  mean_score_time  \
0         0.711528       1.772502         0.009728   
1         0.730228      31.931021         0.009850   
2         0.758627      27.568002         0.012057   
3         0.396193      10.287550         0.031515   
4         0.396193      10.168576         0.031572   
5         0.512147      11.553266         0.031994   
6         0.511292      11.598911         0.031385   

                                        param_models  \
0  LogisticRegression(l1_ratio=1, penalty='elasti...   
1  LogisticRegression(l1_ratio=1, penalty='elasti...   
2  LogisticRegression(l1_ratio=1, penalty='elasti...   
3            DecisionTreeClassifier(random_state=42)   
4            DecisionTreeClassifier(random_state=42)   
5            DecisionTreeClassifier(random_state=42)   
6            DecisionTreeClassifier(random_state=42)   

                                              params  
0  {'models': LogisticRegression(l1_ratio=1, pena...  
1  {'models': LogisticRegression(l1_ratio=1, pena...  
2  {'models': LogisticRegression(l1_ratio=1, pena...  
3  {'models': DecisionTreeClassifier(random_state...  
4  {'models': DecisionTreeClassifier(random_state...  
5  {'models': DecisionTreeClassifier(random_state...  
6  {'models': DecisionTreeClassifier(random_state...

Модель DecisionTreeClassifier показала не очень хорошие результаты. По итогу лучшей моделью выбираем LogisticRegression. Посчитатем метрику на тестовой выборке:

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

        
Кросс-валидация одновремено с извлечением признаков из текстов требует особого подхода. Ты помнишь, что векторайзер мы обучали на обучающей выборке а применяли к тестовой . Что происходит при кросс-валидации? Ты разбиваешь большую обучающую выборку на несколько фолдов, из них выделяешь часть на обучение, часть на валидацию.  Но признаки (TF-IDF) уже обучены на всем наборе (обучающий + валидационный). Получается утечка данных. Чтобы избежать её нужно использовать pipeline из векторайзера и модели.
    
Пример такого пайплайна можно посмотреть в официальной документации
    
  https://scikit-learn.org/stable/auto_examples/model_selection/plot_grid_search_text_feature_extraction.html
    
И еще полезные материалы:
    
  https://newtechaudit.ru/pipeline-v-rabote-s-dannymi/           
        
        
        
</div>

In [15]:
print(f'Метрика RMSE на тестовой выборке: {f1_score(y_test, y_test_pred)}')

Метрика RMSE на тестовой выборке: 0.772386476163531


Метрика получила значение > 0.75, наша модель осуществляет хорошие прогнозы.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
На тестовой выборке получено хорошее качество)

## Выводы

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах.

Мы лемматизировали текст, избавились от слов не имеющих смысловой нагрузки и вычислили TF-IDF, который использовали как входной признак. Целевый признаком был столбец toxic.

Мы обучили разные модели классифицировать комментарии на позитивные и негативные. Которые могут искать токсичные комментарии для отправления их на модерацию.

Лучшей моделью стала LogisticRegression с параметром l1_ratio = 1.

Построенная модель получила значение метрики F1 на тестовой выборке равное 0.77, она точно устроит заказчика.

Как дополнение, можно попробовать рассмотреть другие модели, для поиска лучшей метрики. Также можно получше подобрать параметры для DecisionTreeClassifier, возможно она сможет улучшить результат полученный LogisticRegression.

<div style="border:solid Chocolate 2px; padding: 40px">


<h2> Итоговый комментарий ревьюера <a class="tocSkip"> </h2>    
    
  
У меня сложилось хорошее общее впечатление о проекте, тебе удалось неплохо справиться с этим проектом. Молодец! Подготовлены данные для анализа, изучен каждый параметр. Ппостроено несколько моделей классификации и оценено их качество. Осмысленная аналитика и дельная модельная работа - многое удалось как надо)
    
Отмечу отдельные положительные моменты проекта🙂:
    
- в ходе проекта встречались  функции, помогающие избавиться от дублирования кода;
- хорошая интерпретация;
- при обучении моделей использована кросс-валидация и поиск гиперпараметров.
    
Есть несколько моментов всего, на которые стоит ещё раз взглянуть, я указал их в моих комментариях по ходу проекта. Предлагаю тебе доработать проект по моим комментариям, чтобы довести его до совершенства.

  

    
Если будут вопросы, обращайся, с удовольствием на них отвечу.  
</div>

<div style="border:solid Chocolate 2px; padding: 40px">


<h2> Итоговый комментарий ревьюера 2 <a class="tocSkip"> </h2>    
    
     
  
Теперь почти идеально. Принимаю работу)
    
Если хочешь лучше разобраться в теме, то могу посоветовать тебе: 
    
https://huggingface.co/transformers/model_doc/bert.html
    
https://t.me/renat_alimbekov
    
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM
    
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены
    
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд по трансформеру от создателей pytorch
    
https://transformer.huggingface.co/ - поболтать с трансформером
    
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных методов для трансформеров методов NLP
    
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html
    
    
    
В этом проекте были отработаны все моменты. 
    
Успехов тебе в следующем спринте!
    
Рад был помочь тебе)
    
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны